crates/matrix-sdk/src/room/mod.rs

```diff
@@ -368,6 +368,71 @@ impl Room {
         Ok(())
     }
 
+    /// Leave this room and all predecessors.
+    /// If any room was in [`RoomState::Invited`] state, it'll also be forgotten
+    /// automatically.
+    ///
+    /// Only invited and joined rooms can be left.
+    pub async fn leave(&self) -> Result<()> {
+        let mut rooms: Vec<Room> = vec![self.clone()];
+        let mut cur_room = self;
+
+        'leave_aggregate_loop: loop {
+            let pred = cur_room.predecessor_room();
+            let pred = pred.map(|this_pred| {
+                cur_room
+                    .client
+                    .get_room(&this_pred.room_id)
+                    .ok_or(Error::PredecessorWrongOrForgotten)
+            });
+
+            match pred {
+                None => break 'leave_aggregate_loop,
+                Some(room) => match room {
+                    Ok(room) => {
+                        rooms.push(room);
+                        cur_room = rooms.last().expect("rooms just pushed so can't be empty");
+                    }
+                    Err(e) => {
+                        debug!("Tombstone following terminated early: {e:?}");
+                        break 'leave_aggregate_loop;
+                    }
+                },
+            }
+        }
```

Comment:
I think this whole section could be written like this:

```rust
let mut rooms: Vec<Room> = vec![self.clone()];
let mut current_room = self;

while let Some(predecessor) = current_room.predecessor_room() {
    let maybe_previous_room = current_room.client.get_room(&predecessor.room_id);

    if let Some(previous_room) = maybe_previous_room {
        rooms.push(previous_room.clone());
        current_room = rooms.last().expect("The room was just pushed to the vec");
    }
}
```

This removes the need for the breaks which might be a bit hard to follow since they are effectively gotos.

Reply:
You can add an else branch there that logs if we can't find the previous room.